cmd/ocsp-responder/main.go

```diff
@@ -32,27 +33,101 @@ import (
 	"github.com/letsencrypt/boulder/sa"
 )
 
-/*
-DBSource maps a given Database schema to a CA Key Hash, so we can pick
-from among them when presented with OCSP requests for different certs.
+// ocspFilter stores information needed to filter OCSP requests (to ensure we
+// aren't trying to serve OCSP for certs which aren't ours), and surfaces one
+// boolean method to determine if a given request should be filtered or not.
+type ocspFilter struct {
+	issuerKeyHashAlgorithm crypto.Hash
+	issuerKeyHashes        [][]byte
+	serialPrefixes         []string
+}
+
+// check returns a descriptive error if the request does not satisfy any of
+// the requirements of an OCSP request, or nil if the request should be handled.
+func (f *ocspFilter) check(req *ocsp.Request) error {
+	if req.HashAlgorithm != f.issuerKeyHashAlgorithm {
+		return fmt.Errorf("Request ca key hash using unsupported algorithm %s: %w", req.HashAlgorithm, bocsp.ErrNotFound)
+	}
+	// Check that this request is for the proper CA
+	match := false
+	for _, keyHash := range f.issuerKeyHashes {
+		if match = bytes.Equal(req.IssuerKeyHash, keyHash); match {
+			break
+		}
+	}
+	if !match {
+		return fmt.Errorf("Request intended for wrong issuer cert %s: %w", hex.EncodeToString(req.IssuerKeyHash), bocsp.ErrNotFound)
```

Comment:
Nice use of wrapping! :-D

Reply:
Actually, I looked at the spot where `ErrNotFound` is checked for: https://github.com/letsencrypt/boulder/blob/16c7a21a5773800e15abee14727e42986c27d5f8/ocsp/responder.go#L334

And it uses an equality test, not `errors.Is`, so it will break with this change. We should either switch that to `errors.Is` or return a non-wrapped error here.

Also, the fact that the tests didn't catch this suggests we don't have adequate test coverage. For instance, I think the integration tests need a case that tests for an "unauthorized" response for unknown serials as opposed to an "internal error" response.

Also this is a good reminder that we can't just assume that wrapping an error will do what we expect - we need to check for the places that test for that error's type or value.

Reply:
Good catch; I updated the code which handles this error, but not the code which handles the errors passed up by that code. Got an integration test running, and filed a follow-up bug to fill out a bunch more relevant integration tests.

Reply:
And I confirmed that changing `err = ErrNotFound` to `errors.Is(err, ErrNotFound)` changes the response code from 500 to 401.